packages/dataviews/src/bulk-actions.js

```diff
@@ -21,24 +21,31 @@ const {
 	DropdownMenuSeparatorV2: DropdownMenuSeparator,
 } = unlock( componentsPrivateApis );
 
-function ActionWithModal( { action, selectedItems, setActionWithModal } ) {
+function ActionWithModal( {
+	action,
+	selectedItems,
+	setActionWithModal,
+	bulkActionsButtonRef,
```

Comment:
I don't like this prop. In general I don't like passing refs and Dom elements as props because: Which can of ref is this, an object, a function?... Also, they're subject to timing changes and mutations.

A better API here could be `onPerform` or something like that. Something to be called when the action is done for instance. 

That said, I do wonder: I believe modals and popover have a built-in behavior to restore the focus to the caller (button that opened the popover/modal), in this case it doesn't work natively, so I'm curious to know why the "focus return" don't work by default when the dropdown get closed. cc @ciampo

Reply:
> I believe modals and popover have a built-in behavior to restore the focus to the caller

The `Modal`, when closed, will try to restore focus on the element that was last focused before opening the modal — ie. the menu item. Similarly, the DropdownMenu will try to restore focus on the menu trigger when closed.

Therefore, if we want to focus the dropdown menu trigger when closing the modal, that may be something that needs to be implemented in a custom way. An alternative that we could consider is to leave the dropdown menu open — in theory, the modal component should be able to restore focus correctly on the menu item when closed (see [storybook example](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-dropdownmenu-v2--with-modals)).

Reply:
> The Modal, when closed, will try to restore focus on the element that was last focused before opening the modal — ie. the menu item. Similarly, the DropdownMenu will try to restore focus on the menu trigger when closed.

I guess in my mind, when I click the trigger it should get focused and then open the modal so it's the last item that was focused before the modal is opened. 

But now that you bring this up, I remember some discrepancies between browsers and not all of them have the "focus on click" behavior. Which means that maybe it's because of me testing on Safari which makes me wonder whether we should bother about fixing this issue (should we just revert the commit?) as this is the expected behavior there. I'll check in chrome (before the commit) to confirm.

Reply:
I went ahead and tried to implement an alternative solution that doesn't rely on `bulkActionsButtonRef` at all (and keeps the dropdown menu open in the background when rendering the modal). Could this work?

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/dataviews/src/bulk-actions.js b/packages/dataviews/src/bulk-actions.js
index 9613ff37ee..3fa5db3d76 100644
--- a/packages/dataviews/src/bulk-actions.js
+++ b/packages/dataviews/src/bulk-actions.js
@@ -25,7 +25,7 @@ function ActionWithModal( {
 	action,
 	selectedItems,
 	setActionWithModal,
-	bulkActionsButtonRef,
+	onMenuOpenChange,
 } ) {
 	const eligibleItems = useMemo( () => {
 		return selectedItems.filter( ( item ) => action.isEligible( item ) );
@@ -33,11 +33,10 @@ function ActionWithModal( {
 	const { RenderModal, hideModalHeader } = action;
 	const onCloseModal = useCallback( () => {
 		setActionWithModal( undefined );
-		// Focus the bulk actions button.
-		if ( bulkActionsButtonRef?.current?.focus ) {
-			bulkActionsButtonRef.current.focus();
-		}
-	}, [ setActionWithModal, bulkActionsButtonRef ] );
+		// Close dropdownmenu
+		// TODO: should be called only if the primary action of the modal is clicked
+		onMenuOpenChange( false );
+	}, [ setActionWithModal, onMenuOpenChange ] );
 	return (
 		<Modal
 			title={ ! hideModalHeader && action.label }
@@ -50,31 +49,23 @@ function ActionWithModal( {
 	);
 }
 
-function BulkActionItem( {
-	action,
-	selectedItems,
-	onMenuOpenChange,
-	setActionWithModal,
-	bulkActionsButtonRef,
-} ) {
+function BulkActionItem( { action, selectedItems, setActionWithModal } ) {
 	const eligibleItems = useMemo( () => {
 		return selectedItems.filter( ( item ) => action.isEligible( item ) );
 	}, [ action, selectedItems ] );
+
+	const shouldShowModal = !! action.RenderModal;
+
 	return (
 		<DropdownMenuItem
 			key={ action.id }
 			disabled={ eligibleItems.length === 0 }
-			onClick={ async ( event ) => {
-				event.preventDefault();
-				if ( !! action.RenderModal ) {
-					onMenuOpenChange( false );
+			hideOnClick={ ! shouldShowModal }
+			onClick={ async () => {
+				if ( shouldShowModal ) {
 					setActionWithModal( action );
 				} else {
 					await action.callback( eligibleItems );
-					// Click the bulk actions button to close the dropdown and focus the button.
-					if ( bulkActionsButtonRef?.current?.click ) {
-						bulkActionsButtonRef.current.click();
-					}
 				}
 			} }
 			suffix={
@@ -86,13 +77,7 @@ function BulkActionItem( {
 	);
 }
 
-function ActionsMenuGroup( {
-	actions,
-	selectedItems,
-	onMenuOpenChange,
-	setActionWithModal,
-	bulkActionsButtonRef,
-} ) {
+function ActionsMenuGroup( { actions, selectedItems, setActionWithModal } ) {
 	return (
 		<>
 			<DropdownMenuGroup>
@@ -101,9 +86,7 @@ function ActionsMenuGroup( {
 						key={ action.id }
 						action={ action }
 						selectedItems={ selectedItems }
-						onMenuOpenChange={ onMenuOpenChange }
 						setActionWithModal={ setActionWithModal }
-						bulkActionsButtonRef={ bulkActionsButtonRef }
 					/>
 				) ) }
 			</DropdownMenuGroup>
@@ -131,7 +114,7 @@ export default function BulkActions( {
 			selection.includes( getItemId( item ) )
 		);
 	}, [ selection, data, getItemId ] );
-	const bulkActionsButtonRef = useRef();
+
 	if ( bulkActions.length === 0 ) {
 		return null;
 	}
@@ -143,7 +126,6 @@ export default function BulkActions( {
 				label={ __( 'Filters' ) }
 				trigger={
 					<Button
-						ref={ bulkActionsButtonRef }
 						className="dataviews-bulk-edit-button"
 						__next40pxDefaultSize
 						variant="secondary"
@@ -164,19 +146,14 @@ export default function BulkActions( {
 			>
 				<ActionsMenuGroup
 					actions={ bulkActions }
-					data={ data }
-					selection={ selection }
-					getItemId={ getItemId }
-					onMenuOpenChange={ onMenuOpenChange }
 					setActionWithModal={ setActionWithModal }
 					selectedItems={ selectedItems }
-					bulkActionsButtonRef={ bulkActionsButtonRef }
 				/>
 				<DropdownMenuGroup>
 					<DropdownMenuItem
 						disabled={ areAllSelected }
-						onClick={ ( event ) => {
-							event.preventDefault();
+						hideOnClick={ false }
+						onClick={ () => {
 							onSelectionChange( data );
 						} }
 						suffix={ data.length }
@@ -185,8 +162,8 @@ export default function BulkActions( {
 					</DropdownMenuItem>
 					<DropdownMenuItem
 						disabled={ selection.length === 0 }
-						onClick={ ( event ) => {
-							event.preventDefault();
+						hideOnClick={ false }
+						onClick={ () => {
 							onSelectionChange( [] );
 						} }
 					>
@@ -199,7 +176,7 @@ export default function BulkActions( {
 					action={ actionWithModal }
 					selectedItems={ selectedItems }
 					setActionWithModal={ setActionWithModal }
-					bulkActionsButtonRef={ bulkActionsButtonRef }
+					onMenuOpenChange={ onMenuOpenChange }
 				/>
 			) }
 		</>

```
</details>

Reply:
Actually, I just did more testing and here's my findings:

**First issue**

For the "reset template" action, there's no modal, but what's happening there is that the "dropdown" stays open but the action become "disabled" which means there's a focus loss. I believe there are two things we should do here:

 - The simple solution is to just close the dropdown when the action is done `onActionPerformed` or something which should in theory restore the focus to the trigger.
 - The second thing that we need to do IMO (even if it's probably not necessary once the first solution is implemented) is that "disabled" menu items in the dropdown menu shouldn't be "disabled" entirely, they should use the `aria-disabled` trick.  (Equivalent to the `__experimentalIsFocusable` prop on the Button component). This will allow the buttons to stay focusable and prevent focus loss even though the action is disabled.

**Second issue**

For the "delete template" action, there's a modal, so there are nested dialogs basically. And in this case, everything closes and there's a focus loss. I'm not entirely sure what's happening here so I have no solutions to offer for this case yet, but my guess is that with the "aria-disabled" trick that I suggest above, there's a small chance of this issue being fixed as well.

---

I guess what I'm saying is that we shouldn't try to add custom code to solve these focus loss issues for this particular UI, instead we should try first to address the systemic issues that I raise above. Does that make sense?

Reply:
Related to the first issue, the menu items should not lose focus when becoming disabled as [they are already marked as "accessible when disabled" ](https://github.com/WordPress/gutenberg/blob/ae7467d8bab6cbe538ac0de648b57a62480e2bdc/packages/components/src/dropdown-menu-v2/index.tsx#L57). Have you tried inspecting the `document.activeElement` ?

Related to the second issue, have you tried the solution that I proposed in the [message above](https://github.com/WordPress/gutenberg/pull/57255#discussion_r1442663814)? IMO we should keep the dropdownmenu open under the modal:

- when pressing "cancel" in the modal, we simply close the modal and the focus gets moved back to the dropdown menu automatically;
- when pressing the primary action (ie. "Delete"), we can programmatically set the DropdownMenu open state to `false`, which should then cause the dropdownmenu to automatically move the focus on its trigger.

Reply:
> Have you tried inspecting the document.activeElement and/or using arrow keys to move the selection to other dropdown menu items? Maybe the issue is only visual (ie. the item doesn't look focused because it's disabled).

 - It actually becomes a "div" and not a button anymore. Weird. So it's a similar issue but different trigger. so yes, not the aria-disabled trick.
 - The activeElement is the dropdown wrapper in my tests, so it seems there's some code in place already that handle a focus fallback or something.

Reply:
And for the second issue, you're right as well, the popover is closed when the modal is opened. I'm not entirely sure if we do this on purpose or not but could this be the result of "focus loss". Since the focus is not on the popover anymore, it just closes itself. In other words, our popovers/dropdowns might not support nested dialogs being focused properly yet.

Reply:
> * It actually becomes a "div" and not a button anymore. Weird. So it's a similar issue but different trigger. so yes, not the aria-disabled trick.
> * The activeElement is the dropdown wrapper in my tests, so it seems there's some code in place already that handle a focus fallback or something.

I think all dropdown menu items are actually `div[role="menuitem"]` elements.

Regarding the `activeElement`:
- when using the keyboard to click the "Deselect" item, focus stays on the item
- when using the mouse, focus moves to the menu wrapper

So I believe that focus is already being handled correctly by the underlying ariakit implementation.

> And for the second issue, you're right as well, the popover is closed when the modal is opened. I'm not entirely sure if we do this on purpose or not but could this be the result of "focus loss". Since the focus is not on the popover anymore, it just closes itself.

Closing the dropownmenu when opening the modal is indeed the cause for focus loss. If we kept the dropdownmenu open, the menu item that was clicked to open the modal would receive focus when the modal gets closed.

> In other words, our popovers/dropdowns might not support nested dialogs being focused properly yet.

IMO, the better approach (also UX-wise) is to keep the dropdown menu open under the modal. Doing so should also allow the components to handle focus correctly without extra code:

- Modal would handle focus restoration when dismissing the modal by focusing the correct dropdown menu item
- If we want to force the dropdown menu to close, setting its `open` prop to `false` should move focus automatically on the dropdownmenu trigger. So we should be already able to cover 

When used that way, our components should already handle focus correctly.

Reply:
> IMO, the better approach (also UX-wise) is to keep the dropdown menu open under the modal

I agree but my question was: Is that even possible with our current components. I believe most of our "popovers"... close themselves when they lose focus to a child modal/popover but you say otherwise, so that's news for me :)

Reply:
Yup! The new `DropdownMenu` menu doesn't use our underlying `Modal` component, and can be definitely be kept open under a `Modal`.

Reply:
I applied the patch suggested by @ciampo and implemented the remaining todo specified there. Thank you a lot for providing a detailed patch! Things seem to be working well.
The current behavior is:
- When an action without modal is executed the menu closes and the focus goes to bulk editor button.
- When an action opens a modal the menu keeps open if the modal is closed the focus goes to the button that triggered the modal, if the action in the modal is executed the menu gets closed and the focus goes back to the bulk edit button.

@youknowriad @ciampo  let me know if there is any other update required.